### PR TITLE
feat(server): SSO + RBAC. Closes #3525

### LINKED ISSUE #3525: Role-based Access Controls for SSO
# Summary

Currently any user that logs in using SSO escalates to the `argo-server` service account. We would like other options.

# Motivation

TBD

# Proposal

Either:

* **Casbin style** similar to Argo CD. 
* **Client mode style** i.e. map groups to a service account.

See:

* #1813

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Closes #3525

### CODE REVIEW

**Comment by vladlosev** [comment]
On a side note, this will require `argo-server` to have extra permissions: `list` and `get` on the involved service accounts and `get` on their secrets. Users will have to create extra roles and bindings for `argo-server` with those permissions - or they have to be added to the Helm chart.

**Comment by alexec** [comment]
> On a side note, this will require argo-server to have extra permissions: list and get on the involved service accounts and get on their secrets. 

Yes. I'm not sure why our tests are not highlighting this issue mind you.

File: manifests/quick-start/sso/overlays/argo-server-sa.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+metadata:
+  name: argo-server
+  annotations:
+    # The rule is an expression used to determine if this service account should be used. 
+    # * `groups` - an array of the OIDC groups
+    # * `iss` - the issuer
+    # * `sub` - the subject (typically the username)
+    # Must evaluate to a boolean. 
+    # If you want an account to be the default to use, this rule can be "true".
+    workflows.argoproj.io/rbac-rule: "'authors' in groups"
```
Comment (vladlosev) [inline manifests/quick-start/sso/overlays/argo-server-sa.yaml:12]: Would love some pointers here to the language used for these expressions. My first thought looking here was "How do I link the SA with multiple groups?".

Comment (alexec) [reply manifests/quick-start/sso/overlays/argo-server-sa.yaml:12]: It's this: https://github.com/antonmedv/expr

But yes, we should provide some examples. 

You have:

- `sub`
- `iss`
- `groups` if you have configured them in sso: {groups: []}

Comment (vladlosev) [reply manifests/quick-start/sso/overlays/argo-server-sa.yaml:12]: It then probably makes sense to add something like this:
```suggestion
    # Details of the expression language are available in
    # https://github.com/antonmedv/expr/blob/master/docs/Language-Definition.md.
    workflows.argoproj.io/rbac-rule: "'authors' in groups"
```

File: manifests/quick-start/sso/overlays/argo-server-sa.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    # * `sub` - the subject (typically the username)
+    # Must evaluate to a boolean. 
+    # If you want an account to be the default to use, this rule can be "true".
+    # Details of the expression language are available in
+    # https://github.com/antonmedv/expr/blob/master/docs/Language-Definition.md.
+    workflows.argoproj.io/rbac-rule: "'authors' in groups"
+    # The precedence is used to determine which service account to use whe
+    # Precedence is an integer. It may be negative. If omitted, it defaults to "0".
+    # Numerically higher values have higher precedence (not lower, which maybe counter-intuitive to you).
+    # If two rules match and have the same predence, then which one used will be arbitrary.
```
Comment (tyrken) [inline manifests/quick-start/sso/overlays/argo-server-sa.yaml:18]: If two rules match with same precedence then don't do arbitrary - sort by name or issue an error & don't allow access.  I know not as neat to code but far more understandable to admin-user.

**Comment by vladlosev** [comment]
@alexec have you decided to pursue some other course for implementing the RBAC?

**Comment by alexec** [comment]
No, this is the likely implementation, but we may changes how SSO works underneath.

**Comment by agilgur5** [comment]
Replaced by #4198

**Pull request closed without merging**